store/tikv/gcworker/gc_worker.go

```diff
@@ -511,44 +556,62 @@ func resolveLocks(ctx goctx.Context, store tikv.Storage, safePoint uint64, ident
 	return nil
 }
 
-func doGC(ctx goctx.Context, store tikv.Storage, safePoint uint64, identifier string) error {
-	gcWorkerCounter.WithLabelValues("do_gc").Inc()
-
-	err := saveSafePoint(store.GetSafePointKV(), tikv.GcSavedSafePoint, safePoint)
-	if err != nil {
-		return errors.Trace(err)
-	}
-
-	// Sleep to wait for all other tidb instances update their safepoint cache.
-	time.Sleep(gcSafePointCacheInterval)
+type gcTask struct {
+	startKey  []byte
+	endKey    []byte
+	safePoint uint64
+}
 
-	log.Infof("[gc worker] %s start gc, safePoint: %v.", identifier, safePoint)
-	startTime := time.Now()
-	successRegions := 0
-	failedRegions := 0
+type gcTaskWorker struct {
+	identifier string
+	store      tikv.Storage
+	taskCh     chan *gcTask
+	wg         *sync.WaitGroup
+	// use atomic to read and set
+	successRegions *int32
+	failedRegions  *int32
+}
 
-	ticker := time.NewTicker(gcJobLogTickInterval)
-	defer ticker.Stop()
+func newGCTaskWorker(store tikv.Storage, taskCh chan *gcTask, wg *sync.WaitGroup, identifer string, successRegions *int32, failedRegions *int32) *gcTaskWorker {
+	return &gcTaskWorker{
+		identifer,
+		store,
+		taskCh,
+		wg,
+		successRegions,
+		failedRegions,
+	}
+}
 
-	bo := tikv.NewBackoffer(tikv.GcOneRegionMaxBackoff, ctx)
-	var key []byte
-	for {
-		select {
-		case <-ctx.Done():
-			return errors.New("[gc worker] gc job canceled")
-		case <-ticker.C:
-			log.Infof("[gc worker] %s gc in process, safePoint: %v, successful regions: %v, failed regions: %v, cost time: %s",
-				identifier, safePoint, successRegions, failedRegions, time.Since(startTime))
-		default:
+func (w *gcTaskWorker) run() {
+	defer w.wg.Done()
+	for task := range w.taskCh {
+		err := w.doGCForRange(task.startKey, task.endKey, task.safePoint)
```

Comment:
In region split case, will some worker GC the same region? Is there any problem for this situation?

Reply:
No problem here, see above